File: backend/app/routers/stock.py
from datetime import date

import structlog
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies.auth import get_current_admin, get_current_user
from app.models.stock_movement import MovementType
from app.models.user import User
from app.schemas.stock_movement import (
    StockAdjustRequest,
    StockMovementListResponse,
    StockMovementResponse,
)
from app.services.stock_service import StockService

logger = structlog.get_logger()

router = APIRouter(prefix="/stock", tags=["Stock"])


@router.post(
    "/{product_id}/adjust",
    response_model=StockMovementResponse,
    status_code=status.HTTP_200_OK,
)
async def adjust_stock(
    product_id: str,
    data: StockAdjustRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_admin),
):
    """
    Adjust stock for a product.
    Requires admin. Creates a permanent movement record.
    Publishes a RabbitMQ alert if stock drops to or below threshold.
    """
    try:
        movement = await StockService(db).adjust_stock(
            product_id=product_id,
            data=data,
            adjusted_by=current_user.id,
            adjusted_by_name=current_user.full_name
        )
        return StockMovementResponse.model_validate(movement)


    except ValueError as e:
        error_message = str(e)

        if "not found" in error_message.lower():
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=error_message,
            )
        if "insufficient stock" in error_message.lower():
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=error_message,
            )

        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=error_message,
        )


@router.get(
    "/{product_id}/history",
    response_model=StockMovementListResponse,
    status_code=status.HTTP_200_OK,
)
async def get_movement_history(
    product_id: str,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=10, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Returns paginated movement history for a product.
    Any logged-in user can view history.
    Most recent movements first.
    """
    try:
        return await StockService(db).get_movement_history(
            product_id=product_id,
            page=page,
            page_size=page_size,
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        )
        
@router.get(
    "/movements",
    response_model=StockMovementListResponse,
    status_code=status.HTTP_200_OK,
)

async def get_all_movements(
    movement_type: MovementType | None = Query(default=None),
    start_date: date | None = Query(default=None),
    end_date: date | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=10, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Returns paginated movement history across all products.
    Any logged-in user can view.
    Optional filters: movement_type and/or a date range.
    """
    return await StockService(db).get_all_movements(
        page=page,
        page_size=page_size,
        movement_type=movement_type,
        start_date=start_date,
        end_date=end_date,
    )

File: frontend/src/api/stock.ts
import client from './client'

export type MovementType = 'RECEIVE' | 'SHIP' | 'ADJUST'

export type StockMovement = {
  id: string
  product_id: string | null
  product_sku: string
  movement_type: MovementType
  quantity_change: number
  quantity_before: number
  quantity_after: number
  note: string | null
  created_by: string | null
  created_at: string
}

export type StockMovementListResponse = {
  items: StockMovement[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export type StockAdjustPayload = {
  movement_type: MovementType
  quantity_change: number
  note?: string
}

export const adjustStock = async (
  productId: string,
  data: StockAdjustPayload,
): Promise<StockMovement> => {
  const response = await client.post(`/stock/${productId}/adjust`, data)
  return response.data as StockMovement
}

export const getStockHistory = async (
  productId: string,
  page = 1,
  pageSize = 10,
): Promise<StockMovementListResponse> => {
  const response = await client.get(`/stock/${productId}/history`, {
    params: { page, page_size: pageSize },
  })
  return response.data as StockMovementListResponse
}

export const getAllMovements = async (
  page = 1,
  pageSize = 10,
  movementType?: MovementType,
  startDate?: string,
  endDate?: string,
): Promise<StockMovementListResponse> => {
  const startInstant = startDate
    ? new Date(`${startDate}T00:00:00`).toISOString()
    : undefined
  const endInstant = endDate
    ? new Date(`${endDate}T23:59:59.999`).toISOString()
    : undefined

  const response = await client.get('/stock/movements', {
    params: {
      page,
      page_size: pageSize,
      ...(movementType && { movement_type: movementType }),
      ...(startInstant && { start_date: startInstant }),
      ...(endInstant && { end_date: endInstant }),
    },
  })
  return response.data as StockMovementListResponse
}
